File: java/2015/day04a/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Program {
	public static void usage() {
		System.err.println("usage: java -jar Program.jar <key>");
		System.exit(1);
	}

	public static int process(String key) throws NoSuchAlgorithmException {
		MessageDigest md5 = MessageDigest.getInstance("MD5");
		int n = 1;
		while ( true ) {
			String tryKey = key + String.format("%d", n);
			byte[] digestBytes = md5.digest(tryKey.getBytes());
			StringBuffer digestChars = new StringBuffer();
			for ( byte b : digestBytes ) {
				digestChars.append(String.format("%02x", b));
			}
			String digest = digestChars.toString();
			if ( digest.startsWith("00000") ) {
				break;
			} else {
				n += 1;
			}
		}
		return n;
	}

	public static final void main(String[] args) {
		if ( args.length < 1 ) {
			usage();
		}
		String key = args[0];
		int result = 0;
		try {
			result = process(key);
		} catch ( NoSuchAlgorithmException e ) {
			System.err.println("could not find md5 algorithm:" + e.toString());
			System.exit(1);
		}
		System.out.println("result = " + result);
	}
}

File: kotlin/2015/day04a/Program.kt
import java.security.MessageDigest

fun usage() {
    System.err.println("usage: java -jar Program.jar <input file>")
    kotlin.system.exitProcess(1)
}

fun process(key: String) : Int {
    var n = 1

    while ( true ) {
        val tryKey = "$key$n"
        val digest = MessageDigest.getInstance("MD5").digest(tryKey.toByteArray())
        val hexBytes = digest.joinToString("") { "%02x".format(it) }
        if ( hexBytes.startsWith("00000") ) {
            return n
        } else {
            n += 1
        }
    }
}

fun main(args: Array<String>) {
    if ( args.isEmpty()) {
        usage()
    }
    val key = args[0]
    val result = process(key)
    println("result = $result")
}
